refactor[cartesian]: shiny error messges from the gt_script frontend

romanc opened the pull request:
## Description

Code readability and quality of life refactors around error messages coming from the gtscript frontend.

- Custom `GTError*` classes: added type hints
- Fontend exceptions: cleanup in default messages
- gtscript frontend
  - no `else` needed after `return` or `raise` in `if` statements.
  - leveraging early return to avoid unnecessary indentation levels
  - passing stencil name as `scope` in `nodes.Location` as part of error messages

## Requirements

- [x] All fixes and/or new features come with corresponding tests.
  Covered by existing tests.
- [ ] Important design decisions have been documented in the appropriate ADR inside the [docs/development/ADRs/](docs/development/ADRs/README.md) folder.
  N/A

FlorianDeconinck commented on tests/cartesian_tests/unit_tests/frontend_tests/test_gtscript_frontend.py:
```diff
@@ -823,7 +823,7 @@ def stencil(in_f: gtscript.Field[np.float64]):
         with pytest.raises(
             gt_frontend.GTScriptSyntaxError,
-            match="Cannot nest `with` node inside horizontal region",
+            match="Cannot nest `with` node inside a horizontal region.$",
```
```suggestion
            match="Cannot nest `with` node inside a horizontal region.",
```

FlorianDeconinck reviewed (approved):
LSTM (Looks Shiny To Me)

egparedes commented on src/gt4py/cartesian/definitions.py:
```diff
@@ -131,16 +131,16 @@ class GTError(Exception):
 
 class GTSyntaxError(GTError):
-    def __init__(self, message, *, frontend):
+    def __init__(self, message: str, *, frontend: str):
```
Just a question: did you forget to add the `None` return type hint or you did it on purpose?  I know it's not longer needed so I'm just curious.

egparedes commented on src/gt4py/cartesian/frontend/exceptions.py:
```diff
@@ -6,76 +6,88 @@
 class GTScriptSyntaxError(gt_definitions.GTSyntaxError):
-    def __init__(self, message, *, loc=None):
+    def __init__(self, message: str | None, *, loc: nodes.Location | None = None):
+        if message is None:
```
Optional suggestion in case you also want to generate a proper message if an empty string is passed (it applies to all Exception classes in this file:
```suggestion
        if not message:
```
Additionally, you could implement the logic to add the location info in the base class, as a header or ending of the message passed from the subclasses. It's more work and it probably needs tweaking a bit the error messages so the final string makes sense, so it's up to you to decide, it's just an idea.

egparedes commented on src/gt4py/cartesian/frontend/nodes.py:
```diff
@@ -168,9 +168,10 @@ class Location(Node):
 
     @classmethod
-    def from_ast_node(cls, ast_node, scope="<source>"):
+    def from_ast_node(cls, ast_node, scope: str | None = None):
```
Just a comment: since the default value is immutable, there is nothing wrong with assigning it here and it is more explicit.

Additionally, if you _reaaally_ want to use the same default as in the class init, you could do:
```suggestion
    def from_ast_node(cls, ast_node, scope: str = scope._default):
```

egparedes commented on src/gt4py/cartesian/frontend/gtscript_frontend.py:
```diff
@@ -159,44 +160,44 @@ def _make_axis_bound(
+        elif value is None:
+            LARGE_NUM = 10000
+            seq_name = nodes.Domain.LatLonGrid().sequential_axis.name
+            level = endpt
+            if self.axis_name == seq_name:
                 offset = 0
-            elif isinstance(value, gtscript.AxisIndex):
-                level = nodes.LevelMarker.START if value.index >= 0 else nodes.LevelMarker.END
-                offset = value.index + value.offset
-            elif value is None:
-                LARGE_NUM = 10000
-                seq_name = nodes.Domain.LatLonGrid().sequential_axis.name
-                level = endpt
-                if self.axis_name == seq_name:
-                    offset = 0
-                else:
-                    offset = -LARGE_NUM if level == nodes.LevelMarker.START else LARGE_NUM
             else:
-                raise self.interval_error
+                offset = -LARGE_NUM if level == nodes.LevelMarker.START else LARGE_NUM
+        else:
+            raise self.interval_error
```
This seems like a good place to use pattern matching...

egparedes reviewed (approved):
LGTM. Just optional suggestions and ideas just in case they might be helpful, but feel free to ignore them.

romanc commented:
Found this old branch/PR while cleaning up today. No hurry on this one. I've addressed the feedback and rebased to current mainline. Unless anyone objects I'll just merge once CI back at the end of the week.
